openstack/clustering/v1/profiles/requests.go

```diff
@@ -0,0 +1,42 @@
+package profiles
+
+import (
+	"net/http"
+
+	"github.com/gophercloud/gophercloud"
+)
+
+// CreateOptsBuilder for options used for creating a profile.
+type CreateOptsBuilder interface {
+	ToProfileCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts represents options used for creating a profile
+type CreateOpts struct {
+	Name     string                 `json:"name" required:"true"`
+	Metadata map[string]interface{} `json:"metadata,omitempty"`
+	Spec     map[string]interface{} `json:"spec" required:"true"`
```

Comment:
The Spec is the same between the two. But don't think we should actually define the detail of the Spec because the doc also didn't specify the detail in https://developer.openstack.org/api-ref/clustering/#create-profile ```spec | body | object | The detailed specification of the profile.``` Otherwise we would need to change the code for Spec every time a new k-v pair is introduced for the detailed spec.

Reply:
The `policies.Spec` type currently defines Description, Properties, Type, and Version. Is there a good chance that Senlin is going to add additional fields to this? 

note, this isn't about the values of those fields. For example `Spec.Properties` can continue to be a `map[string]interface{}`. 

Or is there a reason why the `policies` package should have a `Spec` type but the `profiles` package's `Spec` should be different?

/cc @dkt26111

Reply:
@JackKuei is correct that the API docs do not explicitly describe the contents of the Spec field.  If we want gophercloud to strictly mirror the published API docs, the Spec field should not be defined as a structure with detailed fields.

For the `policies` package I decided to list the detailed fields in Spec because the Senlin code enforces that all policies have the `type`, `version`, `description` and `properties` fields: https://github.com/openstack/senlin/blob/master/senlin/policies/base.py#L47

For profile, Senlin code also ensures that all profiles have `type`, `version` and `properties` fields in Spec:
https://github.com/openstack/senlin/blob/master/senlin/profiles/base.py#L47

So I think the `profile` package should have a Spec struct listing those members.  But since `profile` and `policy` have different base classes in the Senlin code, the two packages should not share the same Spec type.

Reply:
Thanks for the info.

>  If we want gophercloud to strictly mirror the published API docs, the Spec field should not be defined as a structure with detailed fields.

This is a blurry area. Gophercloud should interact with the Senlin API correctly. If the docs correctly reflect this, great. Using the Senlin API's code acts as verification to this. Other than verification, the other reason we use the code is so we can implement the correct types where appropriate. Otherwise, all fields everywhere would end up being an `interface{}` (exaggerating :)

@JackKuei is absolutely correct in being mindful of future changes to the fields. In some cases this is something we should consider immediately and in other cases we need to implement a strict type system knowing that additional fields can be added in the future. It's not entirely clear how Gophercloud as a whole can easily manage this. If you add in the scenario of _removing_ fields, then we get into more of a microversion discussion which is a whole other problem.

> For the policies package I decided to list the detailed fields in Spec because the Senlin code enforces that all policies have the type, version, description and properties fields: https://github.com/openstack/senlin/blob/master/senlin/policies/base.py#L47
>
> For profile, Senlin code also ensures that all profiles have type, version and properties fields in Spec:
https://github.com/openstack/senlin/blob/master/senlin/profiles/base.py#L47

This is perfect - thank you :)

> So I think the profile package should have a Spec struct listing those members. But since profile and policy have different base classes in the Senlin code, the two packages should not share the same Spec type.

Agreed. The field name of `Spec` in this case is coincidental. As noted above, the two spec field definitions are slightly different. But since there is a defined schema of some sort in the code, we should create a dedicated type for this.

Thank you both for clearing this up.